src/content/developers/docs/scaling/optimistic-rollups/index.md

```diff
@@ -1,46 +1,254 @@
+
+i. Developers can migrate existing smart contracts on Ethereum to optimistic rollup chains without having to modify codebases extensively. This can save development teams time when deploying Ethereum smart contracts on L2. 
+
+ii. Developers and project teams using optimistic rollups can take advantage of Ethereum's infrastructure. This includes programming languages, code libraries, testing tools, client software, deployment infrastructure, and so on. 
+
+Using existing tooling is important because these tools have been extensively audited, debugged, and improved over the years. It also removes the need for Ethereum developers to learn how to build with an entirely new development stack. 
+
+#### 3. Cross-chain contract calls {#cross-chain-contract-calls} 
+
+Users (externally owned accounts) interact with L2 contracts by submitting a transaction to the rollup contract or having a sequencer or validator do it for them. Optimistic rollups also allow contract accounts on Ethereum to interact with L2 contracts using bridging contracts to relay messages and pass data between L1 and L2. This means you can program an L1 contract on Ethereum Mainnet can to invoke functions belonging to contracts on an L2 ptimistic rollup. 
+
+Cross-chain contract calls happen asynchronously—meaning the call is initiated first, then executed at a later time. This is different from calls between the two contracts on Ethereum, where the call produces results immediately. 
+
+An example of a cross-chain contract call is the token deposit described earlier. A contract on L1 escrows the user's tokens and sends a message to a paired L2 contract to mint an equal amount of tokens on the rollup. 
+
+As cross-chain message calls result in contract execution, the sender is usually required to cover [gas costs](/developers/docs/gas/) for computation. It is advisable to set a high gas limit to prevent the transaction from failing on the target chain. The token bridging scenario is a good example; if the L1 side of the transaction (depositing the tokens) works, but the L2 side (minting new tokens) fails due to low gas, the deposit becomes irrecoverable. 
+
+Finally, we should note that L2 > L1 message calls between contracts need to account for delays (L1 > L2 calls are typically executed after some minutes). This is because messages sent to Mainnet from the optimistic rollup cannot be executed until the challenge window expires.
+
+## How do optimistic rollup fees work? {#how-do-optimistic-rollup-fees-work}
+
+Many optimistic rollups use a “gas price” scheme, much like Ethereum, to denote how much users pay per transaction. How much users pay for transactions on optimistic rollups depends on the following components:
+
+1. **State write**: There is a fixed cost for writing to Ethereum’s state (i.e., submitting a transaction on the Ethereum blockchain). Optimistic rollups reduce this cost by batching transactions and spreading fixed costs across multiple users.  
```

Comment:
It's a cost, but not a fixed one. It depends on the L1 gas price.

Reply:
By "fixed costs", I mean the operator has to pay a fee for sending an on-chain transaction, regardless of how big or small the batch is. How much they pay for calldata is highly variable from what I learned. For context, here's an [article](https://forum.celestia.org/t/ethereum-rollup-call-data-pricing-analysis/141) I read on "variable" vs "fixed " costs for rollups.

Reply:
How about this?

1. **State write**: The minimum cost of an Ethereum transaction is 21k gas plus either 16 or 4 gas per calldata byte (bytes with zeros cost 4 gas, all other values cost 16). Optimistic rollups can reduce the cost of writing the transaction to L1 by batching multiple transactions in a single block (which amortizes the 21k gas over multiple user transactions). They can also compress the transactions to reduce the number of calldata bytes.

Reply:
Thanks for the suggestion. I'd mentioned calldata costs separately from the state write, but combined them based on your suggestion. Here's the new text:

> Optimistic rollups publish transaction data and block headers (consisting of the previous block header hash, state root, batch root) to Ethereum as `calldata`. The minimum cost of an Ethereum transaction is 21,000 gas. Optimistic rollups can reduce the cost of writing the transaction to L1 by batching multiple transactions in a single block (which amortizes the 21k gas over multiple user transactions). 

> Beyond the base transaction fee, the cost of each state write depends on the size of `calldata` posted to L1. Calldata costs are currently governed by [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559), which stipulates a cost of 16 gas for non-zero bytes and 4 gas for zero bytes of calldata, respectively. To reduce user fees, rollup operators compresss transactions to reduce the number of `calldata` bytes published on Ethereum.